tests/trainer/test_fsdp2.py

```diff
@@ -86,3 +87,82 @@ def test_fsdp2_initialization_with_tied_params(
     assert (torch.equal(weight_1, weight_2))
+
+
+@pytest.mark.parametrize('case', ['all_params_one_group', 'subset_one_group', 'multiple_groups'])
+@pytest.mark.parametrize('device', _INIT_DEVICES)
+@world_size(2)
+@pytest.mark.gpu
+@pytest.mark.filterwarnings('ignore:FSDP2 Config/APIs are experimental*:UserWarning')
+@pytest.mark.skipif(SKIP_TEST, reason='FSDP2 is not available in torch < 2.6.0')
+def test_fsdp2_optimizer_handling(
+    case: str,
+    world_size: int,
+    device: str,
+):
+    """Test FSDP2 correctly updates optimizer state for various configurations."""
+    del world_size
+
+    num_classes = 10
+    model = PartialWeightTiedModel(num_features=num_classes, device=device)
+    dataset = RandomClassificationDataset(shape=(num_classes,), size=10, num_classes=num_classes)
+    dataloader = DataLoader(dataset, sampler=dist.get_sampler(dataset))
+
+    all_params_list = list(model.parameters())
+    fc1_params_list = list(model.mlp.fc1.parameters())
+    fc3_params_list = list(model.fc3.parameters())
+    remaining_params_list = [
+        p for p in all_params_list
+        if not any(p is param for param in fc1_params_list)
+        and not any(p is param for param in fc3_params_list)
+        and not any(p is param for param in model.mlp.fc2.parameters()) # To avoid double counting the tied parameter
+    ]
+
+    if case == 'all_params_one_group':
+        optimizer_input = [{'params': all_params_list, 'lr': 0.01}]
+    elif case == 'subset_one_group':
+        optimizer_input = [{'params': fc1_params_list, 'lr': 0.02}]
+    elif case == 'multiple_groups':
+        optimizer_input = [
+            {'params': fc1_params_list, 'lr': 0.01},
+            {'params': fc3_params_list, 'lr': 0.02},
+            {'params': remaining_params_list, 'lr': 0.03},
+        ]
+
+    optimizer = torch.optim.Adam(optimizer_input)
+
+    fsdp2_config = FSDP2Config(
+        device_mesh=None,
+        reshard_after_forward=True,
+        mp_policy=None,
+        offload_policy=None,
+    )
+    prepare_fully_shard(model=model.module, optimizers=optimizer, fsdp2_config=fsdp2_config)
+
+    def validate_optimizer_state(current_optimizer: torch.optim.Optimizer, stage: str):
+        assert len(current_optimizer.param_groups) == len(optimizer_input), \
+            f"[{case}/{stage}] Group count mismatch. Expected {len(optimizer_input)}, Got {len(current_optimizer.param_groups)}"
+        for i, group in enumerate(current_optimizer.param_groups):
+            opt_params = group['params']
+            assert len(opt_params) == len(optimizer_input[i]['params']), \
+                f"[{case}/{stage}] Group {i}: Param count mismatch. Expected {len(optimizer_input[i]['params'])}, Got {len(opt_params)}"
+            assert all(isinstance(p, DTensor) for p in opt_params), \
+                f"[{case}/{stage}] Group {i}: Not all parameters are DTensors"
+            assert group['lr'] == optimizer_input[i]['lr'], \
```

Comment:
should probably check that all keys match? lr might be the only key, but the code has the intention os preserving all so we should check all i think